### PR TITLE
Alerting: Extract values from MD expr alerts

### PR DESCRIPTION
**Pull request by kylebrandt**
**What this PR does / why we need it**:
Extract Eval strings from Multi dimensional alerts.

image

image

Similar to what was done for classic condition: https://github.com/grafana/grafana/pull/34075

In this case, for each labeled number in the selected condition, we look for other labeled _number_ (not time series results) from other queries. If we find one that is a match on labels (exact, or subset) that we add that value with its corresponding refID and labels to the evaluation string. Due to the subset facet (e.g. `host=a,device=lo` and `host=a`) are a match, the labels are included.

**Special notes for your reviewer**:
This is perhaps a bit hacky. It might be better to push this into SSE, and have a call to SEE that knows the target condition (refID) so it can extract this.

But it seems to work and can iterate on this in the future. At least this way the values of things can be seen.

### CODE REVIEW

**Review by papagian — APPROVED** [review]
LGTM

File: pkg/services/ngalert/eval/eval.go
Comment (gotjosh) [inline pkg/services/ngalert/eval/eval.go:183]: is there a chance this could panic? Shouldn't we validate the lenght of `frame.Fields`?

File: pkg/services/ngalert/eval/eval.go
Comment (gotjosh) [inline pkg/services/ngalert/eval/eval.go:172]: Given we're doing comparisons below and from what I can see its only purpose is to attach them to the actual results, I think it would be very beneficial if we started using the fingerprint of these.

I know this type comes from the go sdk, but we can do something like `fingerprint := model.LabelsToSignature(labels.Copy())` here.

File: pkg/services/ngalert/eval/eval.go
Comment (gotjosh) [inline pkg/services/ngalert/eval/eval.go:196]:
```suggestion
        // classic conditions already have metadata set and only have one value, there's no need to add anything in this case.
		if frame.Meta != nil && frame.Meta.Custom != nil {
```
Is this understanding correct? If so, I'd replace the comment below with this one.

File: pkg/services/ngalert/eval/eval.go
Comment (gotjosh) [inline pkg/services/ngalert/eval/eval.go:204]: What happens in other cases? Why don't we attach the data for those?

**Review by gotjosh — COMMENTED** [review]
Great work Kyle 🎉  and thank you very much for patiently waiting on my review. I have a few questions (mostly for the benefit of my understanding) before I approve.

File: pkg/services/ngalert/eval/eval.go
Comment (kylebrandt) [reply pkg/services/ngalert/eval/eval.go:183]: Is done in the continue above.

File: pkg/services/ngalert/eval/eval.go
Comment (kylebrandt) [reply pkg/services/ngalert/eval/eval.go:204]: It would be a time series, or perhaps empty, or something unexpected.

File: pkg/services/ngalert/eval/eval.go
Comment (gotjosh) [inline pkg/services/ngalert/eval/eval.go:177]:
```suggestion
		// for each frame within each response, the response can contain several data types including time-series data.
		// For now, we favour simplicity and only care about single scalar values.
```

**Review by gotjosh — APPROVED** [review]
Those labels will actually make it to the end-user — are you feeling OK with that? I’m a bit weary that if we expose it now and our users start depending on it in its current form we might introduce a breaking change later on.

However, this is experimental after all for a while so I guess we're good to go. 👍

File: pkg/services/ngalert/eval/eval.go
Comment (kylebrandt) [reply pkg/services/ngalert/eval/eval.go:172]: Not sure I fully understand.

Saying replace `theseLabels.Equals(cap.Labels) || theseLabels.Contains(cap.Labels) || cap.Labels.Contains(theseLabels)` with a fingerprint, maybe using the Intersection method?

Comment (gotjosh) [reply pkg/services/ngalert/eval/eval.go:172]: I meant replacing the first equality comparison, do we need the union as well? Under which scenarios would a subset of the labels show up on either side?

**Pull request merged**
